recipes/nsqadmin.rb

```diff
@@ -13,23 +13,62 @@
 nsq_release = "nsq-#{node['nsq']['version']}-#{node['nsq']['go_version']}"
 
 if node['nsq']['setup_services']
-  template '/etc/init/nsqadmin.conf' do
-    action :create
-    source 'upstart.nsqadmin.conf.erb'
-    mode '0644'
-    # need to stop/start in order to reload config
-    if node['nsq']['reload_services']
-      notifies :stop, 'service[nsqadmin]', :immediately
-      notifies :start, 'service[nsqadmin]', :immediately
+
+  if node.platform?('ubuntu') && node['platform_version'].to_f >= 18.04
+
+    template '/etc/systemd/system/nsqadmin.service' do
+      action :create
+      source 'systemd.nsqadmin.conf.erb'
+      mode '0644'
+      notifies :run, 'execute[systemctl-daemon-reload]'
+      # need to stop/start in order to reload config
+      if node['nsq']['reload_services']
+        notifies :stop, 'service[nsqadmin]', :immediately
+        notifies :start, 'service[nsqadmin]', :immediately
+      end
+    end
+
+    template "#{node['chef-nsq']['script_dir']}/nsqadmin-start.sh" do
+      action :create
+      source 'nsqadmin-start.sh.erb'
+      mode '0550'
+      owner node['nsq']['nsqadmin']['user']
+      group node['nsq']['nsqadmin']['user']
+    end
+
+    execute 'systemctl-daemon-reload' do
+      command 'systemctl daemon-reload'
+      action :nothing
+    end
+
+  else
+    template '/etc/init/nsqadmin.conf' do
+      action :create
+      source 'upstart.nsqadmin.conf.erb'
+      mode '0644'
+      # need to stop/start in order to reload config
+      if node['nsq']['reload_services']
+        notifies :stop, 'service[nsqadmin]', :immediately
+        notifies :start, 'service[nsqadmin]', :immediately
+      end
     end
   end
 
   service 'nsqadmin' do
-    provider Chef::Provider::Service::Upstart
-    action [:enable, :start]
-    supports stop: true, start: true, restart: true, status: true
-    if node['nsq']['reload_services']
-      subscribes :restart, "ark[#{nsq_release}]", :delayed
+    if node.platform?('ubuntu') && node['platform_version'].to_f >= 18.04
+      provider Chef::Provider::Service::Systemd
+      action [:enable, :start]
+      supports stop: true, start: true, restart: true, status: true
+      if node['nsq']['reload_services']
+        subscribes :restart, "ark[#{nsq_release}]", :delayed
+      end
+    else
+      provider Chef::Provider::Service::Upstart
+      action [:enable, :start]
+      supports stop: true, start: true, restart: true, status: true
+      if node['nsq']['reload_services']
+        subscribes :restart, "ark[#{nsq_release}]", :delayed
+      end
```

Comment:
It looks like the only thing that is different here is the `provider`?  Should that be the only thing in the if block?

Reply:
Yeah I'll refactor this